### PR TITLE
Added signed multiplier (issue #195), two's-complement conversion funcs, tests for both

### PR DESCRIPTION
**Pull request by nvandervoort**

### CODE REVIEW

File: pyrtl/rtllib/multipliers.py [hunk 1]
```diff
@@ -155,6 +155,44 @@ def tree_multiplier(A, B, reducer=adders.wallace_reducer, adder_func=adders.kogg
...
+    aneg = A[-1]
+    bneg = B[-1]
+    a = pyrtl.WireVector(len(A))
+    b = pyrtl.WireVector(len(B))
+    neg = aneg ^ bneg
+    with pyrtl.conditional_assignment:
+        with aneg:
+            a |= ~A + 1
+        with pyrtl.otherwise:
+            a |= A
```
Comment (Gamrix) [inline pyrtl/rtllib/multipliers.py]: You're duplicating the same 5 lines quite a lot, Could you make that a private function?

Comment (Gamrix) [inline pyrtl/rtllib/multipliers.py]: Make that 6 lines

Comment (nvandervoort) [inline pyrtl/rtllib/multipliers.py]: ^? I see the 5 you're referring to (168-172, for example)

**Comment by Gamrix** [comment]
What's the sixth line that can be abstracted away?

File: pyrtl/rtllib/libutils.py [hunk 2]
```diff
@@ -33,6 +33,39 @@ def str_to_int_array(string, base=16):
...
+    Takes a two's-complement represented value and
+    converts it to a signed integer based on the provided bitwidth.
+    For use with Simulation.inspect() etc. when expecting negative numbers,
+    which it does not recognize
+    """
+    valbl = val.bit_length()
+    if bitwidth < val.bit_length() or val == 2**(bitwidth-1):
+        raise pyrtl.PyrtlError("please choose a larger target bitwidth")
+    if val == 0:
+        return 0
```
Comment (Gamrix) [inline pyrtl/rtllib/libutils.py]: Why does there need to be a special case for zero?

Comment (nvandervoort) [inline pyrtl/rtllib/libutils.py]: That was from the flipping algorithm I was using before, I just removed it

**Comment by nvandervoort** [comment]
Is the sixth line the one declaring the wires (`a`, `b`, and `res`)?

File: pyrtl/rtllib/multipliers.py [hunk 3]
```diff
@@ -172,18 +171,20 @@ def signed_tree_multiplier(A, B, reducer=adders.wallace_reducer, adder_func=adde
             bits[i + j].append(x & y)
 
     res = reducer(bits, bits_length+1, adder_func)
```
Comment (Gamrix) [inline pyrtl/rtllib/multipliers.py]: can you call `tree_mult` instead of making a copy of the code that it had?

Comment (nvandervoort) [inline pyrtl/rtllib/multipliers.py]: It's not the same code, `tree_mult` uses bit_length`, not `bit_length+1`. Also it uses `_trivial_mult()` when one of the inputs is 1 bit, while the signed version is supposed to throw an error there.

Comment (Gamrix) [inline pyrtl/rtllib/multipliers.py]: Why do you need `bit_lengths + 1`? Also, the error is thrown earlier in the function anyways, so having the `trivial_mult` doesn't matter.

Comment (Gamrix) [inline pyrtl/rtllib/multipliers.py]: Also, by the first line of the copied code, we know that `a` and `b` both have a 0 in their most significant bits. shouldn't we shave those off?

Comment (nvandervoort) [inline pyrtl/rtllib/multipliers.py]: Idk, it doesn't work without the `+1`

Comment (nvandervoort) [inline pyrtl/rtllib/multipliers.py]: Oh wait I figured it out, shaved off the first bit, then multiplied, then extended the result and applied `_twos_comp_conditional()` and the result's bitwidth is `len(A) + len(B)`, which is good too

**Pull request merged**